lib/spanner/config/semantic_validator.ex

```diff
@@ -0,0 +1,48 @@
+defmodule Spanner.Config.SemanticValidator do
+
+  alias Piper.Permissions.Ast
+  alias Piper.Permissions.Parser
+
+  @moduledoc """
+  Validates bundle config semantics.
+  """
+
+
+  @doc """
+  Accepts a config map and validates semantics. Returns `:ok` if the config is
+  valid and `{:error, err}` on error.
+  """
+  def validate(config) do
+    verify_rules(config)
+  end
+
+  defp verify_rules(config) do
+    permission_set = MapSet.new(config["permissions"])
+
+    Enum.with_index(config["rules"])
+    |> Enum.reduce([], fn({rule, index}, acc) ->
+      {:ok, %Ast.Rule{}=expr, rule_perms} = Parser.parse(rule)
+      [bundle, command] = String.split(expr.command, ":", parts: 2)
+
+      if config["bundle"]["name"] != bundle do
+        acc = [{"The bundle name '#{bundle}' does not match the name in the bundle", "#/rules/#{index}"}  | acc]
+      end
+      if not(command) in Enum.map(config["commands"], &Map.fetch!(&1, "name")) do
+        acc = [{"The command '#{expr.command}' is not in the list of commands", "#/rules/#{index}"}  | acc]
+      end
+
+      case MapSet.difference(MapSet.new(rule_perms), permission_set) |> MapSet.to_list do
+        [] ->
+          acc
+        bad_perms ->
+          Enum.map(bad_perms, fn(bad_perm) ->
+            {"The permission '#{bad_perm}' is not in the list of permissions.", "#/rules/#{index}"}
+          end) ++ acc
+      end
+    end)
+    |> (fn ([]) -> :ok
+           (errors) -> {:error, {:bad_rules, errors}}
+        end).()
+  end
```

Comment:
Was having a little bit of trouble following this. Here are a couple ideas to break this down:

* Using pipelines across functions with multiple lines is pretty tough to read. I'd recommend breaking down each piece of the pipeline into it's own function so it's really easy to see the flow of data.
* I think you can actually use `Enum.flat_map` rather than `Enum.reduce` here which should make it easier to follow state accumulation.
* Breaking out each rule error into it's own function would hide a lot of details and make it easy to follow visually.

Here's a quick untested example that incorporates the above:

```
def validate(config) do
  rules       = Map.fetch!(config, "rules")
  bundle      = Map.fetch!(config, "bundle")
  command     = Map.fetch!(config, "command")
  permissions = Map.fetch!(config, "permissions")

  errors = rules
  |> Enum.with_index
  |> Enum.flat_map(&validate_rule(&1, bundle, command, permissions)

  case errors do
    [] ->
      :ok
    errors ->
      {:errors, {:bad_rules, errors}}
  end
end

def validate_rule({rule, index}, bundle, command, permissions) do
  {:ok, %Ast.Rule{}=expr, rule_permissions} = Parser.parse(rule)
  [rule_bundle, rule_command] = String.split(expr.command, ":", parts: 2)

  errors = [validate_bundle(rule_bundle, bundle),
            validate_command(rule_command, commands),
            validate_permissions(rule_permissions, permissions)]

  errors
  |> Enum.map(&{&1, "#/rules/#{index}"})
  |> List.flatten
end

def validate_bundle(bundle, bundle, index),
  do: []
def validate_bundle(rule_bundle, bundle, index),
  do: ["The bundle name '#{bundle}' does not match the name in the bundle"]

def validate_command(rule_command, commands) do
  command_names = Enum.map(commands, &Map.fetch!(&1, "name")

  case command in command_names do
    true ->
      []
    false ->
      ["The command '#{command}' is not in the list of commands"]
  end
end

def validate_permissions(rule_permission, permissions) do
  missing_permissions = MapSet.difference(MapSet.new(rule_permissions), Map.set(permissions)) |> MapSet.to_list

  Enum.map(missing_permissions, fn permission ->
    "The permission '#{permission}' is not in the list of permissions."
  end)
end
```

And plus you could easily write some tests for each individual validator.